rclpy/rclpy/node.py

```diff
@@ -460,7 +460,16 @@ def declare_parameters(
                 if not descriptor.dynamic_typing and value is not None:
                     # infer type from default value
                     if not isinstance(value, ParameterValue):
-                        descriptor.type = Parameter.Type.from_parameter_value(value).value
+                        deduced_type = Parameter.Type.from_parameter_value(value).value
+                        if descriptor.type != ParameterType.PARAMETER_NOT_SET:
```

Comment:
`descriptor.type` and `descriptor.name` are ignored by all `declare_parameter()` methods.
The stored descriptor is updated with the actual name and type.

I don't think that approach makes much sense, but that's how it has always worked.

If a parameter of an specified type needs to be declared without a default value, you can use:

```python3
node.declare_parameter('name', Parameter.Type.MY_PARAM_TYPE)
```

Reply:
So do you propose to also add a check for the name or not checking for inconsistencies at all? 
I think the test case I fixed shows nicely that we should use all information given by the user to prevent surprises.

Reply:
IMO, either of them looks good to me.

- throw exception since user input is inconsistent. (this PR suggestion)
- actual name and type will be honored against descriptor. (current behavior, now i am inclined to this one, see https://github.com/ros2/rclpy/issues/912#issuecomment-1069681520) but i would suggest that it would be better to add doc for this behavior to avoid the surprise.

i will keep this thread open to hear from others.

> The stored descriptor is updated with the actual name and type.

this behavior is same with `rclcpp`. (so if we change this, `rclcpp` should be changed accordingly.)

Reply:
I'll extend the PR so that we also check for a name-mismatch and throw an exception.

Reply:
>  how about below?

That sounds fine to me